ADR/0009-pipeline-service-via-operator.md

```diff
@@ -0,0 +1,40 @@
+# 0009 Pipeline Service via Operator
+
+Date: 2022-12-01
+
+## Status
+
+Proposed
+
+## Context
+
+kcp is no longer being used as a control plane for StoneSoup. This means that
+"Pipeline Service" cannot be deployed as an independent service. For our
+initial MVP, all Tekton APIs need to be deployed onto a standard OpenShift
+cluster (specifically OpenShift Dedicated).
+
+## Decision
+
+All Tekton APIs will be provided using the stock OpenShift Pipelines operator.
+In the spirit of developing in a "service first" manner, StoneSoup will deploy
+a candidate "nightly" release of the operator. The service will be defined in
```

Comment:
that's good for testing environments but for production?

Reply:
We'll want to employ some level of quality gates on our staging environment before we release to production. We must aspire to merge production-grade code, or we will fail in this endeavor. The time from code merge to code deploy _in production_ needs to get to 4 hours, not 4 days (or 4 months, which is the time it takes for new features to be released today).

Reply:
> all of which are developed by, possibly, a different set of people with a different set of rules,

We'll need to influence the projects/repos to have CI tests which give us production-like quality such that a merge-to-main commit is deployable with reasonable guarantees. 

> So when we talk about nightly here, is it about “nightly” code from the pipeline service, from the additional, owned feature of OpenShift Pipelines or from upstream components ?

In my opinion, 

Long-term, everything.
Short-term, (a) The core Tekton APIs could move slow and steady (b) The non-core Tekton APIs could be cutting edge.

Reply:
> We'll need to influence the projects/repos to have CI tests which give us production-like quality such that a merge-to-main commit is deployable with reasonable guarantees.

Even so, we should not necessarily be relying entirely on upstream CI as that can easily be changed outside of our control. We should still work to have our own quality gates in place, even if that is as minimal as ensuring that the content of the checks hasn't been degraded and that the tests passed on the commits (i.e. not force pushed/merged).

Reply:
It looks like currently, we have a [Subscription](https://github.com/redhat-appstudio/infra-deployments/blob/main/components/pipeline-service/openshift-pipelines/subscription.yaml) to the redhat product version of openshift pipelines. The proposal is to use a different `source` for that Subscription, and a different `channel`.

The channel on its own is problematic, because each night when the nightly build is produced - I think OLM would automatically deploy it, without quality checks on our end.

If we can configure the CatalogSource in infra-deployments and pin that to an image digest, that will give us an ability to deliberately increment to the next nightly build, forcing our CI to run before rolling it out.